File: algorithms/java/fun_all_combination_sum_to_target_with_size_k/Combinations.java
class Combinations {
    
    // easier to understand way
    public List<List<Integer>> combinations1(int[] array, int target, int k) {
        List<List<Integer>> res = new ArrayList<>();
        if (array == null || array.length == 0 || k <= 0) return res;
        Arrays.sort(array);
        List<Integer> cur = new ArrayList<>();
        dfs1(array, target, k, res, cur, 0);
        return res;
    }

    private void dfs1(int[] array, int target, int k, List<List<Integer>> res, List<Integer> cur, int index) {
        if (index == array.length || target < 0 || cur.size() > k) return;
        if (target == 0 && cur.size() == k) {
            res.add(new ArrayList<>(cur));
            return;
        }
        cur.add(array[index]);
        dfs(array, target - array[index], k, cur, index + 1);
        cur.remove(cur.size() - 1);
        while (index < array.length - 1 && array[index] == array[index + 1]) {
            index++;
        }
        dfs(array, target, k, res, cur, index + 1);
    }

    // a better way: call less using iteration to avoid using recursion to go to the next level
    public List<List<Integer>> combinations2(int[] array, int target, int k) {
        List<List<Integer>> res = new ArrayList<>();
        if (array == null || array.length == 0 || k <= 0) return res;
        Arrays.sort(array);
        List<Integer> cur = new ArrayList<>();
        dfs(array, target, k, res, cur, 0);
        return res;
    }

    public void dfs1(int[] array, int target, int k, List<List<Integer>> res, List<Integer> cur, int index) {
        if (index == array.length || target < 0 || cur.size() > k) return;
        if (target == 0 && cur.size() == k) {
            res.add(new ArrayList<>());
            return;
        }
        for (int i = index, i < array.length; i++) {
            if (i == index || array[i] != array[i - 1]) {
                cur.add(array[i]);
                dfs(array, target - array[i], k, res, cur, i + 1);
                cur.remove(cur.size() - 1);
            }
        }
    }
}

File: algorithms/cpp/77_combinations/Combinations.cpp
class Combinations {
public:
	// 1.
	vector<vector<int>> combine1(int n, int k) {
        vector<vector<int>> res;
        vector<int> cur;
        dfs1(res, cur, n, k);
        return res;
    }
    void dfs1(vector<vector<int>>& res, vector<int>& cur, int n, int k) {
        if (cur.size() == k) {
            res.push_back(cur);
            return;
        }
        if (n == 0) {
            return;
        }
        dfs1(res, cur, n - 1, k);
        cur.push_back(n);
        dfs1(res, cur, n - 1, k);
        cur.pop_back();
    }

	// 2.
	vector<vector<int>> combine2(int n, int k) {
        vector<vector<int>> res;
        vector<int> cur;
        dfs2(res, cur, n, k, 1);
        return res;
    }
    void dfs2(vector<vector<int>>& res, vector<int>& cur, int n, int k, int index) {
        if (cur.size() == k) {
            res.push_back(cur);
            return;
        }
        for (int i = index; i <= n; ++i) {
            cur.push_back(i);
            dfs2(res, cur, n, k, i + 1);
            cur.pop_back();
        }
    }
};
